Added diagnostic filter for global code check.

aqez opened the pull request:
Adds support for #504 like so:
```
let g:OmniSharp_diagnostic_exclude_paths = [
\ 'obj\\',
\ 'Temp\\',
\ '\.nuget\\'
\]
```

Before:
![image](https://user-images.githubusercontent.com/8353400/86408992-aac20f80-bc7d-11ea-9fb4-5aa145945806.png)

After:
![image](https://user-images.githubusercontent.com/8353400/86409003-b1e91d80-bc7d-11ea-8d8c-7c3b73973aaa.png)

nickspoons commented:
Thanks for picking this up!

Is there a particular reason for handling the exclusions in `diagnostics.vim`, instead of in `OmniSharp#locations$Parse()`? That's where the `g:OmniSharp_diagnostic_overrides` are already handled, it makes sense to me to add these exclusions in there too.

Also, would you be able to document the new variable in `doc/omnisharp-vim.txt`, and a brief mention in `README.md`?

aqez commented:
My pleasure!

I put it in diagnostics.vim instead because I wasn't sure if OmniSharp#locations$Parse() was used in from other parts of the code (sounded generic). I can definitely move it there if you think it makes more sense. 

I will certainly add to the documentation and readme. Will pick back up on this tomorrow though I'm done for the day! :) Haha

nickspoons commented:
It _is_ generic, and it's also used for e.g. `:OmniSharpGotoDefinition` and `:OmniSharpFindUsages`, but the diagnostic handling/filtering only occurs when there's a `LogLevel` present: [locations.vim#50](https://github.com/OmniSharp/omnisharp-vim/blob/master/autoload/OmniSharp/locations.vim#L50)

It could be argued that it makes more sense to handle this part separately in `diagnostics.vim`, since it is exclusively diagnostic-related. On the other hand, the idea behind the `Parse()` function was to handle all conversions of server locations to Vim "quickfix" objects in one place. Plus, the looping through quickfixes already happens there so it saves having to do that twice.

One way or the other, I'd prefer to have `g:OmniSharp_diagnostic_overrides` and `g:OmniSharp_diagnostic_exclude_paths` handled in the same place, since they are so clearly related. Maybe we should not use `OmniSharp#locations#Parse` at all for diagnostics, but duplicate that functionality in `diagnostics.vim` `s:StdioCheckRH()`? Then we can simplify `OmniSharp#locations#Parse()` significantly, removing all the `g:OmniSharp_diagnostic_overrides` and `g:OmniSharp_diagnostic_showid` logic.

aqez commented:
I added a callback to be called per quickfix inside `locations.vim` `OmniSharp#locations#Parse`. This callback can return 0 or the quickfix or a modified quickfix. Using this from `diagnostics.vim` we can now filter out and modify quickfixes using `g:OmniSharp_diagnostic_exclude_paths` and `g:OmniSharp_diagnostic_overrides`. I also added documentation to the vim doc and `README.md` Let me know what you think! I'm open to other ideas on how to handle this or any other changes you'd like.

aqez commented on autoload/OmniSharp/actions/diagnostics.vim:
```diff
@@ -56,7 +56,41 @@ endfunction
+    endfor
+  endif
+
+  let loglevel = get(a:quickfix, 'LogLevel', '')
+  if loglevel !=# ''
```
This should probably just be `has_key(a:quickfix, 'LogLevel')` since it's not used anywhere.

aqez replied:
Actually.. can this be removed altogether?  @nickspoons

nickspoons replied:
Actually I think everything from [locations.vim 57-79](https://github.com/OmniSharp/omnisharp-vim/pull/606/files#diff-de967f1765544fcf25b39ee9228a0b91R57-R79) can be moved in here - then we don't need loglevel there but we do need it for diagnostics.

nickspoons commented on autoload/OmniSharp/locations.vim:
```diff
@@ -23,14 +23,20 @@ function! OmniSharp#locations#Navigate(location, noautocmds) abort
-    let text = get(quickfix, 'Text', get(quickfix, 'Message', ''))
-    if get(g:, 'OmniSharp_diagnostic_showid') && has_key(quickfix, 'Id')
-      let text = quickfix.Id . ': ' . text
+    let quickfix = Fixup(quickfix)
```
I don't think we need the default function, what about this:

```vim
let quickfix = a:0 ? a:1(quickfix) : quickfix
```

aqez replied:
I'm really new to vimscript (this is my first ever writing it) so this is super helpful thanks! I made the proposed change and it works.

nickspoons replied:
And you're already passing around funcref callbacks 😄 Great stuff. The language definitely has its quirks but I quite like it.

nickspoons commented:
Closes #504 

Thanks @aqez, this very useful and much appreciated!

nickspoons commented:
Hey @aqez I ended up refactoring the location/diagnostics stuff a bit more - I removed the diagnostic callback and no longer use `OmniSharp#location#Parse()` at all for diagnostics. This keeps the location parsing very light and simple.

I did this because I added in another diagnostic location [here](https://github.com/OmniSharp/omnisharp-vim/blob/master/autoload/OmniSharp/stdio.vim#L76) where unsolicited diagnostics are received from the server, and diagnostics from here weren't getting filtered correctly.

This comment is just to give you some context if you suddenly find your code changed 😄

aqez commented:
This seems to have broken it for me - commit a7ea2dc works for me, but when swapping to master (0b6da5f) it breaks.

nickspoons commented:
Oh man, yeah I stuffed it up didn't I. I'll just fix that up...

nickspoons commented:
Yeah that was dumb, and I clearly didn't test it properly 😅. Fixed now.
